### ISSUE TITLE
[OpenCV] Build against conan-gstreamer

### ISSUE THREAD

**Issue by mattdibi** [conan-community/community#204]
### Description of Problem, Request, or Question
I would like to add the GStreamer support to OpenCV (`WITH_GSTREAMER` option) . I've seen that GStreamer [was recently added](https://github.com/bincrafters/conan-gstreamer) to the available packages and would like to use it for conan-opencv.
Is it possible? What steps should I take?

**Comment by SSE4** [comment]
@mattdibi sure, I'll try to add required changes and let you know.

**Comment by mattdibi** [comment]
@SSE4 Awesome! Thank you :D

**Comment by SSE4** [comment]
blocked by https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/issues/605

**Comment by mattdibi** [comment]
@SSE4 Thanks for the update! I'll keep an eye on the issue to follow the development.

**Comment by mattdibi** [comment]
@SSE4 Hi, any update on this?

**Comment by SSE4** [comment]
unfortunately, I've stuck with GStreamer for a while, because of some conan limitations. I am waiting for the conan components feature to be implemented.

**Comment by SSE4** [comment]
@mattdibi I've managed to finish conan packages for `gstreamer` and `gst-plugins-base`. I've added initial GStreamer support to the OpenCV: https://github.com/conan-community/conan-opencv/commit/5195fcbd0d8dcb56b461b0bf5decf5bc9abec2a2
please give a try and lemme know if it works for you

**Comment by mattdibi** [comment]
You guys are awesome! I'll give it a try ASAP.

**Comment by mattdibi** [comment]
Hi,
the `conan install` command complete successfully, but when I later try to compile and build my project against OpenCV I get a problem with missing Protobuf lib.

During the OpenCV build i get the following:

```
opencv/4.1.0@conan/stable: Calling package()
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Compiler GCC>=5, checking major version 5
-- Conan: Checking correct version: 5
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Adjusting fPIC flag (ON)
-- Conan: C++ stdlib: libstdc++11
-- Looking for ccache - not found
-- Found OpenEXR: /home/pcnv2build/.conan/data/openexr/2.3.0/conan/stable/package/9895e0a97c9c932ae8613b72ce9cfb0e9dc93e39/lib/libIlmImf-2_3_s.a
-- Checking for module 'gtk+-3.0'
--   No package 'gtk+-3.0' found
-- Checking for module 'gtk+-2.0'
--   No package 'gtk+-2.0' found
-- Checking for module 'gthread-2.0'
--   Package 'libpcre', required by 'glib-2.0', not found
-- Could not find OpenBLAS include. Turning OpenBLAS_FOUND off
-- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off
-- Could NOT find Atlas (missing: Atlas_CBLAS_INCLUDE_DIR Atlas_CLAPACK_INCLUDE_DIR Atlas_CBLAS_LIBRARY Atlas_BLAS_LIBRARY Atlas_LAPACK_LIBRARY) 
-- Could NOT find BLAS (missing: BLAS_LIBRARIES) 
-- LAPACK requires BLAS
-- A library with LAPACK API not found. Please specify library location.
CMake Warning at /usr/local/share/cmake-3.13/Modules/FindProtobuf.cmake:495 (message):
  Protobuf compiler version doesn't match library version 3.5.1
Call Stack (most recent call first):
  source_subfolder/cmake/OpenCVFindProtobuf.cmake:26 (find_package)
  source_subfolder/CMakeLists.txt:775 (include)


-- Could NOT find JNI (missing: JAVA_AWT_LIBRARY JAVA_JVM_LIBRARY JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 JAVA_AWT_INCLUDE_PATH) 
-- VTK is not found. Please set -DVTK_DIR in CMake to VTK build directory, or to VTK install subdirectory with VTKConfig.cmake file
-- ADE: Download: v0.1.1d.zip
```

**Comment by SSE4** [comment]
seems like GLib wasn't linked, should be easy to fix. I'll try to do that locally and let you know.

**Comment by SSE4** [comment]
@mattdibi I've applied required fixes: https://github.com/conan-community/conan-opencv/commit/606b04af961fcb9197a5c261f59b31a68d58380e
please give a new try and lemme know

**Comment by mattdibi** [comment]
Great! Thank you. I'll give it a try tomorrow.

**Comment by mattdibi** [comment]
@SSE4 Still experiencing some issues during linking the same .so object.

```
[ 18%] Linking CXX executable ../bin/UTILS
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_module_open"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_query_latency"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_fullinfo"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_submit_buffer_list"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_set_format"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_base64_encode"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_sink_wait_preroll"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_is_empty"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_thread_pool_set_max_unused_threads"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_free"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_thread_pool_push"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_double"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_audio_format_from_string"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_audio_format_build_integer"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_pointer"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_sint64"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_thread_pool_new"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_new"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_get_type"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "glib_micro_version"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_uri_unescape_string"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_pop_head"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_void"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_get_stringtable_entries"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_get_stringnumber"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_peek_tail"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_module_error"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "glib_minor_version"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_audio_format_to_string"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_get_do_timestamp"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_thread_pool_free"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_sint32"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_sink_get_type"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_config"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_compile2"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_uint64"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_call"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_module_close"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_free"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_is_live"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_push_tail"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_set_live"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "glib_major_version"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_module_make_resident"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_audio_get_channel_reorder_map"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "g_module_symbol"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_exec"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_queue_array_pop_tail"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "ffi_type_uint32"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_audio_format_get_info"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "gst_base_src_set_caps"
/home/build/.conan/data/opencv/4.1.0/conan/stable/package/5ff6910c0640a28e9a652ba1c85ff7b3dc61a5f5/lib/libopencv_videoio.so.4.1.0: undefined reference "pcre_study"
... etc etc
```

**Comment by mattdibi** [comment]
@SSE4 Hi, I gave it a go with the following `conanfile.txt`:

```
[requires]
boost/1.66.0@conan/stable
gtest/1.8.1@bincrafters/stable
opencv/4.1.0@conan/stable

[generators]
cmake

[options]
boost:shared=True
opencv:gstreamer=True
opencv:protobuf=False
```

It builds successfully but it doesn't seem to find the gstreamer base plugins:

```
0:00:00.002225917  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "appsrc"!
0:00:00.002283762  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "appsrc"
0:00:00.002316156  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "videoconvert"!
0:00:00.002392658  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "videoconvert"
0:00:00.002413553  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002435140  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "x264enc"!
0:00:00.002454961  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "x264enc"
0:00:00.002475254  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002539420  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "rtph264pay"!
0:00:00.002564249  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "rtph264pay"
0:00:00.002585340  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002609651  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "tee"!
0:00:00.002629389  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "tee"
0:00:00.002648035  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002668313  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "queue"!
0:00:00.002685448  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "queue"
0:00:00.002705674  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002723816  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "udpsink"!
0:00:00.002742000  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "udpsink"
0:00:00.002761096  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no source [sink=@(nil)]
0:00:00.002781800  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "udpsink"!
0:00:00.002801905  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:816:priv_gst_parse_yyparse: no element "udpsink"
0:00:00.002821321  4398      0x25fa430 ERROR           GST_PIPELINE grammar.y:901:priv_gst_parse_yyparse: link has no sink [source=t@(nil)]
0:00:00.002999660  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "encodebin"!

(EXE:4398): GLib-GObject-CRITICAL **: 16:24:02.823: g_object_set: assertion 'G_IS_OBJECT (object)' failed
0:00:00.003126266  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "appsrc"!
0:00:00.003233852  4398      0x25fa430 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:gst_element_factory_make: no such element factory "filesink"!
```

I understand that it's missing `x264enc` which is from the ugly plugins but the `appsrc` and `videoconvert` are from the base plugins which is installed.

**Comment by SSE4** [comment]
hi @mattdibi 
it's static `gstreamer` and static `gst-plugins-base`, right?
I suppose in order to use static GStreamer plugins they have to be explicitly initialized by application, right? I mean according to the [doc](https://gstreamer.freedesktop.org/documentation/gstreamer/gstplugin.html?gi-language=c), you have to set `GST_PLUGIN_STATIC_DECLARE(coreelements)` / `GST_PLUGIN_STATIC_REGISTER(coreelements)`, otherwise plugins aren't available

**Comment by SSE4** [comment]
also, have a look at our examples:
https://github.com/bincrafters/conan-gstreamer/blob/testing/1.16.0/test_package/test_package.cpp#L8
https://github.com/bincrafters/conan-gst-plugins-base/blob/testing/1.16.0/test_package/test_package.cpp#L8

**Comment by SSE4** [comment]
otherwise, you may also set environment variable `GST_PLUGINS_PATH` to point to your GStreamer plugins (such as x264enc)

**Comment by SSE4** [comment]
I am closing as changes were applied. please open new issue, if needed.
